src/client/EventRouter.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @hidden
+ */
+
+import {EventEmitter} from 'events';
+
+import {EventSpecification} from '../provider/model/APIHandler';
+
+import {TransportMappings, TransportMemberMappings} from './internal';
+
+type EmitterProvider = (targetId: string) => EventEmitter;
+type EventDeserializer<E extends EventSpecification, T extends E> = (event: EventTransport<E, T>) => E;
+
+interface EventTarget {
+    type: string;
+    id: string;
+}
+export type Targeted<T extends EventSpecification> = T & {
+    /**
+     * If present, will be used to find the correct emitter.
+     *
+     * Allows events to be raised from client-side objects that mirror a corresponding provider-side object. If target
+     * is omitted, events will be raised from a top-level/"global" event emitter.
+     */
+    target?: EventTarget;
+}
+
+export type Transport<E extends EventSpecification, T extends E> = TransportMappings<T> extends never ? {
+    [K in keyof T]: TransportMemberMappings<T[K]>;
+} : TransportMappings<T>;
+export type EventTransport<E extends EventSpecification, T extends E> = Targeted<Transport<E, T>>;
+
+/**
+  * Class for helping take events that have arrived at the client via the IAB channel, and dispatching them on the correct client-side object
+  */
+export class EventRouter<E extends EventSpecification> {
```

Comment:
@bryangaleOF There's been a few changes here from the version in FDC3 (both in this file, and also in where this gets used), would appreciate some input.